### PR TITLE
Pubsub refactor Phase 1 of 3

### PR DESCRIPTION
**Pull request by JackyWYX**
## Issue

This is the phase one for [PubSub refactor](https://github.com/harmony-one/harmony/issues/3322). Implement a new module `harmony/p2p/pubsub` to provide PubSub interface. 

1. This PR refactors the current code logic we are using with libp2p PubSub. 
2. All business logics can be implemented within the `Handler` interface, and be decoupled according to handler's usage. 
3. The handlers can be dynamically registered, added, stopped and removed.

Followings are some public interfaces that will give a taste about how to use this module.

```
type PubSub interface {
	AddHandler(psh Handler) error
	StartHandler(spec HandlerSpecifier) error
	StopHandler(spec HandlerSpecifier) error
	RemoveHandler(spec HandlerSpecifier) error
	RemoveTopic(topic Topic) error
	SendMessageToTopic(ctx context.Context, topic Topic, msg []byte) error
	Start()
	Close()
}
```

Supported handlers:

```
type Handler interface {
	Topic() Topic
	Specifier() HandlerSpecifier
	ValidateMsg(ctx context.Context, peer PeerID, rawData []byte) ValidateResult
	DeliverMsg(ctx context.Context, rawData []byte, cache ValidateCache)
	Report()
}
```

### Unit Test Coverage

```
PASS
coverage: 81.4% of statements
```

Also two race tests are implemented to detect race conditions. 

* `TestTopicRunner_race`
* `TestPubSubHost_Race`

Since race tests will take 10 seconds to finish and we do not have `--race` flag in our test suite, these two tests are skipped for now. To run these two tests, simply to comment out `t.Skip` line and run test with `--race`flag.

### CODE REVIEW

File: p2p/pubsub/psHost.go [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+			return false
+		}
+	}
+	return true
+}
+
+func (psh *pubSubHost) deleteHandlersWithTopic(topic Topic) {
+	for spec, handler := range psh.handlers {
+		if handler.Topic() == topic {
+			delete(psh.handlers, spec)
```
Comment (rlan35) [inline p2p/pubsub/psHost.go:316]: These are run in single threaded code, so no race condition right?

Comment (JackyWYX) [reply p2p/pubsub/psHost.go:316]: Yes. All updates on the running handlers are done in a single thread. You can see the logic in `pubSubHost.run`

File: p2p/pubsub/topicRunner.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+	return
+}
+
+func (tr *topicRunner) validateMsg(ctx context.Context, peer PeerID, raw *libp2p_pubsub.Message) libp2p_pubsub.ValidationResult {
+	m := newMessage(raw)
+	handlers := tr.getHandlers()
+	vResults := make([]ValidateResult, 0, len(handlers))
+
+	for _, handler := range handlers {
```
Comment (LeoHChen) [inline p2p/pubsub/topicRunner.go:93]: are you expecting multiple validator for one message? This can be low efficient as validator has to run on every single libp2p messages.

File: p2p/pubsub/metric.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
+package pubsub
```
Comment (LeoHChen) [inline p2p/pubsub/metric.go:1]: if we are designing a new metrics system, we'd better use a global metrics system that can be added for each layer.  So that we don't have to add metrics system for each layer.

Comment (LeoHChen) [reply p2p/pubsub/metric.go:1]: I am thinking to add a new metrics system using Prometheus https://prometheus.io/docs/guides/go-application/

**Comment by LeoHChen** [comment]
if we are going to redesign the network layer, we should re-consider to have stream support as well as the pubsub.  So that the node can deliver specific message to certain node for streaming, while also deliver broadcast message using pubsub for consensus.

File: p2p/pubsub/psHost.go [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+		}
+	}
+}
+
+// Close close the pubSubHost.
+func (psh *pubSubHost) Close() {
+	task := closeTask{closed: make(chan struct{})}
+	psh.closeTaskC <- task
+
+	<-task.closed
```
Comment (rlan35) [inline p2p/pubsub/psHost.go:119]: why is this needed?

File: p2p/pubsub/topicRunner.go (hunk 2 above)
Comment (JackyWYX) [reply p2p/pubsub/topicRunner.go:93]: Yes, there will be node handler validating node message and consensus handler handling consensus message. So basically, messages are only validated by one handler - Another one is always return `MsgAccept`

File: p2p/pubsub/psHost.go (hunk 4 above)
Comment (JackyWYX) [reply p2p/pubsub/psHost.go:119]: This is to wait for the close task to be executed at `run`.

File: p2p/pubsub/metric.go (hunk 3 above)
Comment (JackyWYX) [reply p2p/pubsub/metric.go:1]: Oh, yea. This can be much greater than the simple metric we implement ourselves. Let me think it over and maybe we can do in another thread.

**Comment by JackyWYX** [comment]
> if we are going to redesign the network layer, we should re-consider to have stream support as well as the pubsub. So that the node can deliver specific message to certain node for streaming, while also deliver broadcast message using pubsub for consensus.

Totally agree. This is definitely what I have been thinking of. Will implement later.

**Review by rlan35 — APPROVED** [review]
lgtm.

File: p2p/pubsub/topicRunner.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+	handlers := tr.getHandlers()
+
+	for _, handler := range handlers {
+		tr.deliverMessageForHandler(ctx, msg, handler)
+	}
+}
+
+func (tr *topicRunner) deliverMessageForHandler(ctx context.Context, msg *message, handler Handler) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
```
Comment (JackyWYX) [inline p2p/pubsub/topicRunner.go:130]: Wrong usage of cancel. Take care of this.
